### PR TITLE
Add Bidirectional Search Implementation

### PR DESCRIPTION
**Pull request by Manthan-R-Sheth**
The PR is aimed at implementing Bidirectional Search for the AIMA4e build.

### CODE REVIEW

**Comment by Manthan-R-Sheth** [comment]
I shall be writing comments in the code and unit tests once the code is good to go.

**Comment by Manthan-R-Sheth** [comment]
@ctjoreilly  if you could review this implementation.

**Comment by ctjoreilly** [comment]
This overlaps with pull request #242 , so I need to look at both together.

**Comment by ctjoreilly** [comment]
Also, could you include a Unit test and fully document your implemtation.

Thanks

Ciaran

**Comment by Manthan-R-Sheth** [comment]
@ctjoreilly  ya i will do so and update the PR.
Thanks.

**Comment by Manthan-R-Sheth** [comment]
@ctjoreilly i have updated the PR, if you can review and give your feedback. 
I have added unit test, and have extended BasicProblem with BidirectionalProblem to improve its functionality and allowing it to be generic as well.
Thanks.

File: core/src/main/java/aima/core/environment/support/ProblemFactory.java
Comment (globalworming) [inline core/src/main/java/aima/core/environment/support/ProblemFactory.java]: start imports: unnecessary change and bad practice IMHO
http://stackoverflow.com/a/147461/1359903
I'd revert that

File: test/src/test/java/aima/test/unit/search/uninformed/BidirectionalSearchTest.java
Comment (globalworming) [inline test/src/test/java/aima/test/unit/search/uninformed/BidirectionalSearchTest.java]: unittest tests 2 things, should be split into "start is goal" test and "from sibiu to fagaras" test

File: core/src/main/java/aima/core/environment/support/ProblemFactory.java
Comment (globalworming) [inline core/src/main/java/aima/core/environment/support/ProblemFactory.java]: naming: this method does not return a BiDirectionalSearch as the name suggests but a BasicProblem

**Review by globalworming — CHANGES_REQUESTED** [review]
unittests could test more cases e.g. zerind->sibiu and arad->neamt and arad->sibiu
use concrete imports instead of star imports

File: test/src/test/java/aima/test/unit/search/uninformed/BidirectionalSearchTest.java
Comment (Manthan-R-Sheth) [reply test/src/test/java/aima/test/unit/search/uninformed/BidirectionalSearchTest.java]: I though to do so, but according to the previous practise followed as seen [here](https://github.com/aimacode/aima-java/blob/AIMA4e/test/src/test/java/aima/test/unit/search/uninformed/BreadthFirstSearchTest.java), I combined the two tests into one.
@globalworming

**Comment by Manthan-R-Sheth** [comment]
@globalworming 
Updated the PR.

File: core/src/main/java/aima/core/search/basic/support/BasicProblem.java
Comment (globalworming) [inline core/src/main/java/aima/core/search/basic/support/BasicProblem.java]: Please Check your other changes for Star imports

Comment (Manthan-R-Sheth) [reply core/src/main/java/aima/core/search/basic/support/BasicProblem.java]: Fixed the star imports. @globalworming, 
Actually Alt+Enter of IntelliJ does that. I will take care henceforth.

**Comment by ctjoreilly** [comment]
I am having problems applying your patch, due to the multiple commits. Did you do any force commits to your pull request? Could I ask you to collapse all your changes into a single squashed pull request?

**Comment by Manthan-R-Sheth** [comment]
@ctjoreilly
I will rebase and squash my commits and update the PR.

**Comment by Manthan-R-Sheth** [comment]
@ctjoreilly
I have squashed my commits and updated the PR.

File: core/src/main/java/aima/core/environment/support/ProblemFactory.java
Comment (ctjoreilly) [inline core/src/main/java/aima/core/environment/support/ProblemFactory.java]: goalLocations -> goalLocation, i.e. should be singular.

File: core/src/main/java/aima/core/search/basic/support/BasicProblem.java
Comment (ctjoreilly) [inline core/src/main/java/aima/core/search/basic/support/BasicProblem.java]: More efficient to instantiate just once and keep as an attribute (see earlier comments).

File: core/src/main/java/aima/core/search/api/BidirectionalProblem.java
Comment (ctjoreilly) [inline core/src/main/java/aima/core/search/api/BidirectionalProblem.java]: Change to '@author manthan'

File: core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java
Comment (ctjoreilly) [inline core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java]: TODO - we need to come up with a way to avoid this dirty cast, suggestions?

Comment (ctjoreilly) [reply core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java]: For now I would suggest adding a SearchForActionsBidirectionallyFunction interface similar to SearchForActionsFunction but that takes a BidirectionalProblem. Update all related code appropriately.

File: core/src/main/java/aima/core/environment/support/ProblemFactory.java
Comment (ctjoreilly) [inline core/src/main/java/aima/core/environment/support/ProblemFactory.java]: Should return a BidirectionalProblem.

File: core/src/main/java/aima/core/search/api/BidirectionalProblem.java
Comment (ctjoreilly) [inline core/src/main/java/aima/core/search/api/BidirectionalProblem.java]: Remove blank line.

File: core/src/main/java/aima/core/search/basic/support/BasicProblem.java
Comment (ctjoreilly) [inline core/src/main/java/aima/core/search/basic/support/BasicProblem.java]: Declare implements BidirecationalProblem after Problem<A, S>, i.e.:

implements Problem<A, S>, Bidirectional....

so that is matches the order these interfaces are introduced in the book.

File: core/src/main/java/aima/core/search/basic/support/BasicProblem.java
Comment (ctjoreilly) [inline core/src/main/java/aima/core/search/basic/support/BasicProblem.java]: change this to:

private Problem<A, S> reverseProblem and instantiate in the Bidirectional constructor, so that it only needs to be done so once.

File: core/src/main/java/aima/core/search/basic/support/BasicProblem.java
Comment (ctjoreilly) [inline core/src/main/java/aima/core/search/basic/support/BasicProblem.java]: Move this to below the problem constructors (and ensure you have a blank line between method definitions). Again so that the order matches that of the book in terms of introduction to concepts.

File: core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java
Comment (ctjoreilly) [inline core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java]: Change to '@author manthan'

File: core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java
Comment (ctjoreilly) [inline core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java:60]: Is there a cleaner way to do this without repeating the conditional block twice?

File: core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java
Comment (ctjoreilly) [inline core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java]: Minor tidy up, change to:
//
// Supporting Code

File: core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java
Comment (ctjoreilly) [inline core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java]: May want to consider a more efficient way of doing this. Have a look at BasicFrontierQueue, it will let you do a containment check much faster.

File: test/src/test/java/aima/test/unit/search/uninformed/BidirectionalSearchTest.java
Comment (ctjoreilly) [inline test/src/test/java/aima/test/unit/search/uninformed/BidirectionalSearchTest.java]: // NOTE: The answer should include the goal as well (i.e. Bucharest) otherwise the list of actions doesn't actually get you there. This should be fixed. I would also consider adding some additional tests.

**Review by ctjoreilly — CHANGES_REQUESTED** [review]
Hi, 

This looks good as an initial version of a core implementation of the BidirectionalSearch algorithm. Could you please go through the review comments and make the appropriate changes and the re-push (again all squashed together as I apply as a patch).

Thanks

Ciaran

File: core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java
Comment (Manthan-R-Sheth) [reply core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java]: @ctjoreilly I will have to change ```NodeFactory<A,S>``` for making a new child node since now the problem is ```BidirectionalProblem``` and not Problem
In my opinion, ```Node<A, S> newChildNode(Problem<A, S> problem, Node<A, S> parent, A action)``` should besuch that it includes both types of problems.

Comment (Manthan-R-Sheth) [reply core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java]: Can I have ```Node<A, S> newChildNodeForBidirectionalSearch(BidirectionalProblem<A, S> problem, Node<A, S> parent, A action)``` as a function of ```NodeFactory<A,S>``` ?

File: core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java
Comment (Manthan-R-Sheth) [reply core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java]: ```front``` and ```back``` can be made ```BasicFrontierQueue<A,S>``` . It is good way for improving efficiency.

File: core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java
Comment (Manthan-R-Sheth) [reply core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java:60]: I am not sure about an alternative way for this. This is because alternate evaluations of both queues need to be done for the algorithm.

File: test/src/test/java/aima/test/unit/search/uninformed/BidirectionalSearchTest.java
Comment (Manthan-R-Sheth) [reply test/src/test/java/aima/test/unit/search/uninformed/BidirectionalSearchTest.java]: Since the output is a list of Actions ( ```List<A>``` ), the goal has a null action and so I didn't add it to the result.

**Comment by Manthan-R-Sheth** [comment]
@ctjoreilly if you could review now. 
**Changes made** :- 
1. SearchForActionsBidirectionallyFunction is implemented that takes BidirectionalProblem.
1. newChildNodeForBidirectionalSearch is added to make nodes for bidirectional search.
1. Queues are converted to BasicFrontierQueue for efficient state containment. (can be extended to other classes too).
1. reverseProblem attribute is added in BasicProblem.
1. Unit tests are increased.
1. Minor fixes of @author and new lines.

File: core/src/main/java/aima/core/search/api/BidirectionalProblem.java
Comment (ctjoreilly) [inline core/src/main/java/aima/core/search/api/BidirectionalProblem.java]: Why are you doing the above? Ideally, I'd like to keep this the way it was. I would like to also avoid changing the NodeFactory to take this into account could you explain further?

**Review by ctjoreilly — CHANGES_REQUESTED** [review]
See review comment related to BidirectionalProblem interface. Also, this pull request has jumped to 13 files. Could you create a new pull request that only conains your changes (and not those also pulled from the aima4e branch).

Thanks

File: core/src/main/java/aima/core/search/api/BidirectionalProblem.java
Comment (Manthan-R-Sheth) [reply core/src/main/java/aima/core/search/api/BidirectionalProblem.java]: @ctjoreilly 
I did this because :-

1. newChildNode() function requires a ```Problem```, but we have ```BidirectionalProblem``` as the argument of apply(). Hence to create a node for such case, I modified it accordingly. Could you suggest a better approach.
1. If BidirectionalProblem isn't implemented as above, we won't have functions like ```initialState()```. Thus to make it just like other types of problems like ```Problem```,
``` NondeterministicProblem``` etc I did so.

**Comment by Manthan-R-Sheth** [comment]
@ctjoreilly  I have rebased onto my commits and git history is clean.

File: core/src/main/java/aima/core/search/basic/support/BasicProblem.java
Comment (ctjoreilly) [inline core/src/main/java/aima/core/search/basic/support/BasicProblem.java]: You shouldn't need to track the finalState if you've already instantiated the reverseProblem as I don't see where it will be used otherwise.

File: test/src/test/java/aima/test/unit/search/uninformed/BidirectionalSearchTest.java
Comment (ctjoreilly) [inline test/src/test/java/aima/test/unit/search/uninformed/BidirectionalSearchTest.java]: Thinking about this some more, I think we may need to change the interface on SearchForActionsBidirectionallyFunction to return something other than a List<A>. Don't consider addressing this until we get a basic version merged into the trunk that we agree upon and then we can come back round to see if this idea make sense:

Instead it returns something like this:

 interface BidirectionalActions<A> {
      List<A> fromInitialStatePart()
      List<A> fromGoalStatePart()

      List<A> fromInitialStateToGoalState(); // What I was indicating above I was expecting as  a result
      List<A> fromGoalStateToInitialState(); // essential the reverse of the previous method.
}

File: test/src/test/java/aima/test/unit/search/uninformed/BidirectionalSearchTest.java
Comment (ctjoreilly) [reply test/src/test/java/aima/test/unit/search/uninformed/BidirectionalSearchTest.java]: Thinking about this some more, I think we may need to change the interface on SearchForActionsBidirectionallyFunction to return something other than a List<A>. Don't consider addressing this until we get a basic version merged into the trunk that we agree upon and then we can come back round to see if this idea make sense:

Instead it returns something like this:

 interface BidirectionalActions<A> {

      List<A> fromInitialStatePart()
      List<A> fromGoalStatePart()
      List<A> fromInitialStateToGoalState(); // What I was indicating above I was expecting as  a result
      List<A> fromGoalStateToInitialState(); // essential the reverse of the previous method.
}

Comment (Manthan-R-Sheth) [reply test/src/test/java/aima/test/unit/search/uninformed/BidirectionalSearchTest.java]: Ya i will surely update it when things get finalised.

File: core/src/main/java/aima/core/search/basic/support/BasicProblem.java
Comment (Manthan-R-Sheth) [reply core/src/main/java/aima/core/search/basic/support/BasicProblem.java]: Actually it used to initialise the reverse problem , in the constructor.
```		this.reverseProblem = new BasicProblem<>(finalState, actionsFn, resultFn, goalTestPredicate, stepCostFn);```
It acts as the initial state for the reverse problem.
I will remove the global variabe though.

**Comment by Manthan-R-Sheth** [comment]
@ctjoreilly  Now is it good to go.

File: core/src/main/java/aima/core/search/api/BidirectionalProblem.java
Comment (ctjoreilly) [reply core/src/main/java/aima/core/search/api/BidirectionalProblem.java]: I thought I'd submitted a response as regards this earlier. I do not wish to go with the BidirectionalProblem API in this form or require any changes to the NodeFactory. Could you explain to me the difference between the results from call BidrectionalProblem.initialState() and BidirectionalProblem.getOriginalProblem().initialState()? Similarly, for the other overlapping methods. Basically, I only see BidirectionalProblem needing to be a container for the original and reverse versions of the problem, it itself does not need to be one. Let me know if I'm missing something here.

Thanks

Ciaran

Comment (Manthan-R-Sheth) [reply core/src/main/java/aima/core/search/api/BidirectionalProblem.java]: BidirectionalProblem.initialState() and BidirectionalProblem.getOriginalProblem().initialState(), both will have same results. I made BidirectionalProblem.getOriginalProblem() for this purpose only. But I used it as a different problem to keep continuity with the superclasses of BasicProblem (which are all problems in themselves). 
Shall I replace BidirectionalProblem.getOriginalProblem().initialState() ? Your call on this aspect.

@ctjoreilly

**Comment by ctjoreilly** [comment]
Hi,

To avoid confusion with the semantic meaning of BidireactionalProblem I think the simplest thing to do is delete it and revert BasicProblem and NodeFactory back to the way they were. Instead modify the apply() method on SearchForActionsBidirectionallyFunction to:

List<A> apply(Problem<A, S> originalProblem, Problem<A, S> reverseProblem);

and the method on ProblemFactory should now return:

Pair<Problem<GoAction, InState>, Problem<GoAction, InState>> getSimpleBidirectionalSearchProblem(...)

This will also help with reducing API bloat. Once done and all tests passing I think this should be good for merging to AIMA4e as the basis for further improvements on BidirectionalSearch.

Thanks

Ciaran

File: core/src/main/java/aima/core/search/api/SearchForActionsBidirectionallyFunction.java
Comment (ctjoreilly) [inline core/src/main/java/aima/core/search/api/SearchForActionsBidirectionallyFunction.java]: This should not take a pair, but broken apart as:

originalProblem, reverseProblem

Using Pair on the ProblemFactory was for convenience as you can't specificy generics in an Array i.e. Problem[].

**Comment by Manthan-R-Sheth** [comment]
@ctjoreilly  Reverted back with BasicProblem and NodeFactory.
Updated accordingly.

**Comment by ctjoreilly** [comment]
Could you squash/rebase so I can validate the patch before accepting the pull request (almost there :) ).

Thanks

Ciaran

**Comment by Manthan-R-Sheth** [comment]
Actually I had rebased locally earlier but forgot to push it on my fork. And I made changes on it locally. I had to merge after that. I cleaned that history by cherrypicking but now I am not able to squash my commits. 
if you could test my PR locally by ```git fetch origin pull/ID/head:BRANCHNAME``` instead of applying patch.
Thanks.

**Comment by ctjoreilly** [comment]
Ok, I'll chance a direct merge as I think it should be good (hopefully)

**Comment by Manthan-R-Sheth** [comment]
@ctjoreilly 
Thanks.

**Pull request merged**
